### PR TITLE
Add no_std support

### LINKED ISSUE #201: no_std support
Let's make a tracking issue for this

### PR DESCRIPTION
**Pull request by mcroad**
Closes #201. Related to bitcoindevkit/bdk#205

Not ready. 

- [x] `no_std` example
- [x] updated readme with instructions
- [x] tests pass
- [x] rust-bitcoin/rust-bitcoin#637 includes `secp256k1/alloc` feature in `bitcoin/no-std`. Released on `0.28.0`
- [x] rust-bitcoin/rust-bitcoin#690 Rust `1.47` set as `no_std` MSRV. Released in `0.28.1`.

Maintains backward compatibility. To use it in a `no_std` context set `default-features = false` and enable the `no-std` feature.

~~To the maintainers: I added the `no-std-compat` crate. It wraps `core` and `hashbrown` and makes them look like `std`. This is a different than what `rust-bitcoin` did. They use `core` and `hashbrown` directly. I believe using `no-std-compat` makes the code more readable. I'd like to hear your thoughts on this.~~ Now using `hashbrown` and `core` directly.

### CODE REVIEW

**Comment by mcroad** [comment]
I need help getting the rust 1.29.0 tests to pass. Specific test here: https://github.com/mcroad/rust-miniscript/runs/4061570991.

```
Run ./contrib/test.sh
+ FEATURES=compiler rand use-serde
+ [ -n  ]
+ [  = true ]
+ [  = true ]
+ cargo test --verbose
 Downloading bitcoin v0.27.1
 Downloading no-std-compat v0.4.1
error: unable to get packages from source

Caused by:
  failed to parse manifest at `/home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/no-std-compat-0.4.1/Cargo.toml`

Caused by:
  editions are unstable

Caused by:
  feature `edition` is required

this Cargo does not support nightly features, but if you
switch to nightly channel you can add
`cargo-features = ["edition"]` to enable this feature
Error: Process completed with exit code 101.
```
I'm not familiar enough with this version of rust to debug it.

**Comment by sanket1729** [comment]
I think this is because `no-std-compat` does not support rust 1.29 :(. I am not well versed with embedded code, maybe that is the reason why rust-bitcoin uses the underlying crates directly?

**Comment by mcroad** [comment]
Removed the `no-std-compat`. All test pass. Waiting on rust-bitcoin/rust-bitcoin#637 to be released and for an answer regarding `no_std` MSRV rust-bitcoin/rust-bitcoin#690.

Please approve CI workflow.

**Comment by sanket1729** [comment]
Sorry, Github is annoying for first-time contributors.

**Comment by apoelstra** [comment]
Approved the github workflow.

**Comment by mcroad** [comment]
Last 2 items on the checklist are done. Will squash and rebase on master.

**Comment by mcroad** [comment]
@apoelstra Given `rust-bitcoin` moved it's MSRV to `1.41.1`, will `rust-miniscript` follow suit?

**Comment by sanket1729** [comment]
@mcroad, we just moved to 2018 and this PR now has lots of conflicts. Sorry about that.

**Comment by mcroad** [comment]
Just rebased. Still working on the PR. It's not ready for review.

**Comment by mcroad** [comment]
All test pass but PR will not be final until rust-bitcoin/rust-bitcoin#690 is released.

https://github.com/mcroad/rust-miniscript/actions/runs/2261662000

Until then, I await your comments on the PR as it is.

File: src/lib.rs [hunk 1]
```diff
@@ -99,6 +100,23 @@
 #![deny(missing_docs)]
 
 pub use bitcoin;
+
+#[cfg(not(any(
+    any(feature = "std", feature = "no-std"),
+    all(feature = "std", feature = "no-std")
+)))]
+compile_error!("at least one of the `std` or `no-std` features must be enabled, but not both");
```
Comment (apoelstra) [inline src/lib.rs:108]: In f8855d172471b56ef17d8f248f3c029f8895905b:

This makes me uncomfortable. Are you able to make this work such that `std` just overrides `no-std`? For that matter, are you able to eliminate the `no-std` feature and just do nostd stuff whenever `std` is not enabled?

Comment (apoelstra) [reply src/lib.rs:108]: Grepping around I actually don't see where the `no-std` feature is even used.

Comment (mcroad) [reply src/lib.rs:108]: I agree that having both `std` and `no-std` features is weird. Best practice is to have a single `std` feature that's enabled by default that turns `std` usage on. The [Rust book agrees](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification). They specifically say:

>For example, if you want to optionally support [`no_std`](https://doc.rust-lang.org/reference/names/preludes.html#the-no_std-attribute) environments, **do not** use a `no_std` feature. Instead, use a `std` feature that *enables* `std`.

The main issue is that the only way to enable `rust-bitcoin`'s `no_std` mode, is to enable and re-export its `no-std` feature.

I would also rather have a single `std` feature. This is why I never reference `no-std` outside of the Cargo config and `src/lib.rs`. My hope is that eventually `rust-bitcoin` will adopt best practice and remove the `no-std` feature.

---
This is somewhat related to rust-bitcoin/rust-bitcoin#758.

Comment (apoelstra) [reply src/lib.rs:108]: Ah, I see, I didn't realize that rust-bitcoin was similarly broken. Let's see if we can fix that now that we have a msrv bump..

**Comment by sanket1729** [comment]
Sorry about this, but this would face a lot of conflicts because of our move to 2018. We can focus on getting `no-std` in rust-miniscript now that it is released in rust-bitcoin.

**Comment by mcroad** [comment]
Rebased and finally got tests to pass.

File: src/lib.rs (hunk 1 above)
Comment (apoelstra) [reply src/lib.rs:99]: What if you removed this line and just set `no_std = ["bitcoin/no_std"]` in Cargo.toml?

I agree that it's ugly but at least it won't break compilation when users enable features.

**Comment by sanket1729** [comment]
@mcroad Really sorry again for the conflicts. I will put this on top of my review list so that there are no more conflicts after this one

File: src/lib.rs (hunk 1 above)
Comment (mcroad) [reply src/lib.rs]: Removed the compiler error.

**Comment by mcroad** [comment]
Rebased again.

**Review by apoelstra — APPROVED** [review]
ACK 28d304e88c58e0595643b2c60402c4839d92a28e

I'm a little sad to see that we can no longer compile without default features, but I guess that's been true of rust-bitcoin for a long time, so c'est la vie.

**Review by sanket1729 — COMMENTED** [review]
Thanks for the PR. We can address the comments in a follow PR if required. 

I have zero expirience in no-std development. I just tried to follow the instructions to build the no-std build. Initially, I could not get it to build and tried to mimic the CI. Installing the `rust-src` component seemed to do the trick. Should we also that in the README?

I ran into a linker error:

> sanket1729:~/rust-miniscript/embedded$ echo $RUSTFLAGS 
-C link-arg=-Tlink.x

> sanket1729:~/rust-miniscript/embedded$ echo $CARGO_TARGET_THUMBV7M_NONE_EABI_RUNNER 
qemu-system-arm -cpu cortex-m3 -machine mps2-an385 -nographic -semihosting-config enable=on,target=native -kernel

> sanket1729:~/rust-miniscript/embedded$ cargo run +nightly --target thumbv7m-none-eabi
   Compiling typenum v1.15.0
   Compiling cortex-m v0.7.4
   Compiling proc-macro2 v1.0.39
   Compiling libc v0.2.126
   Compiling memchr v2.5.0
   Compiling syn v1.0.95
   Compiling cortex-m-rt v0.6.15
   Compiling cortex-m v0.6.7
error: linking with `cc` failed: exit status: 1
  = note: "cc" "-m64" "/tmp/rustcu8JyXT/symbols.o" "/home/sanket1729/rust-miniscript/embedded/target/debug/build/cortex-m-78cb07deae82bed4/build_script_build-78cb07deae82bed4.build_script_build.2257045c-cgu.0.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/debug/build/cortex-m-78cb07deae82bed4/build_script_build-78cb07deae82bed4.build_script_build.2257045c-cgu.1.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/debug/build/cortex-m-78cb07deae82bed4/build_script_build-78cb07deae82bed4.build_script_build.2257045c-cgu.10.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/debug/build/cortex-m-78cb07deae82bed4/build_script_build-78cb07deae82bed4.build_script_build.2257045c-cgu.11.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/debug/build/cortex-m-78cb07deae82bed4/build_script_build-78cb07deae82bed4.build_script_build.2257045c-cgu.12.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/debug/build/cortex-m-78cb07deae82bed4/build_script_build-78cb07deae82bed4.build
  
 Probably, something else might be off in my system because the following works :shrug: 
 > RUSTFLAGS=$RUSTFLAGS CARGO_TARGET_THUMBV7M_NONE_EABI_RUNNER=$CARGO_TARGET_THUMBV7M_NONE_EABI_RUNNER cargo run --target thumbv7m-none-eabi --release
    Finished release [optimized + debuginfo] target(s) in 0.02s

   
2) I am also not able to build this in debug mode (without the --release flag). Is this expected? 

> Compiling embedded v0.1.0 (/home/sanket1729/rust-miniscript/embedded)
error: linking with `rust-lld` failed: exit status: 1
  |
  = note: "rust-lld" "-flavor" "gnu" "/tmp/rustc90zJr1/symbols.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1a4vfh3jry9qhmtf.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1aoq3x7wmhn8n29u.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1b8rm12f45qaldm2.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1h6zyj1rmnpmfcoi.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1itggcb0fccfedej.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1k8nddveuc69crrg.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1kuhmdd5hcuxpir8.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1lc2im5g0e0lu66d.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1nam0vrwxtu0s903.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1pzhzutagmqjg64z.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1u7uty6l4hxvj89p.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1u9ph96gmb8vymbh.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.1uel84j51kktbvcm.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.23c0uin0qhigctw.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.24eq9zezcp1zrsmt.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.251oqgzfi9t5az0y.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.2a6qgbkabaaclpvc.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.2j7gbbp2u9sykfgo.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.2k8irx23kmk13rk6.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.2r9whzo63erlzygp.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.2x525qruyaeln3nr.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.3b5k6z5sphn7j3pl.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.3bv8lgwtl141o5pj.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.3m75df4qbs5wjltp.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.3n3eeumeiyxvb2mo.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.3nkoszhj0sxtzc94.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.3qjzqm0l32uclc5z.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.3s0ohddi4j5jayga.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.3sh0l8v86dqb6k25.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.439wmgcig5q70hpt.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.456gzcon6482dqlt.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.48ycblph89kak52b.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.4dkbcmqoteiqjb9r.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.4eob32fyhrzpmg0a.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.4ernbojsis0t20rg.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.4hnu9bhv3y9syh5g.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.4o58r3nx75v3a8b2.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.4oqndp04mbr4zwx6.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.4t6q0gvit5vmtqh2.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.4v9i8wqsssropkr7.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.4z58clcdwpuno69g.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.512qk1ak5bimee9p.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.51aupj3hj8733ih0.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.5b5q4lecv8en8ted.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.5buru3ykyz9m97sh.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.5cjlfujuitjcc6cv.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.cgvhk5ocpbte80m.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.egkotks9bo7ddef.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.g51hkg8i617m5wd.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.gk60olq4xfsfvq6.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.ijujnka561m924c.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.xzuqetkl9uhq6et.rcgu.o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788.3c6nn0krnharmlvm.rcgu.o" "--as-needed" "-L" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps" "-L" "/home/sanket1729/rust-miniscript/embedded/target/debug/deps" "-L" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/build/cortex-m-aeea63aea6ffdc2f/out" "-L" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/build/cortex-m-rt-076aecb144ce0d41/out" "-L" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/build/cortex-m-semihosting-a68a8a89ca250f7f/out" "-L" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/build/secp256k1-sys-1d469d5bff022d46/out" "-L" "/home/sanket1729/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/thumbv7m-none-eabi/lib" "-Bstatic" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libminiscript-bc8f8ca73da25eca.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libbitcoin-ec03fc4a63c88d70.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libhashbrown-2288f4321118a88f.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libahash-6cc44301a08b6cd7.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libbech32-990ecd751650535e.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libsecp256k1-ad59a94bc87e11e8.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libsecp256k1_sys-c03b8d37363bd6ad.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libbitcoin_hashes-16d8139e1876d7d7.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libcore2-b53703a46090b77a.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libhashbrown-9c86b8600d94aec9.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libahash-2e4b8b19a0c5ef7f.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libcortex_m_semihosting-424ad78302277fc2.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libcortex_m_rt-367afd52f356e9af.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libr0-777d267c3f5e883a.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libcortex_m-58c69b609c826e0c.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libaligned-efb924f7f569df32.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libas_slice-7cc634893bc17dca.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libstable_deref_trait-659fd4fc938e6d63.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libgeneric_array-f32408a42fd0c608.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libgeneric_array-48c0a4d35713388a.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libgeneric_array-2acafc3aedb05a8a.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libtypenum-6293bcb1e7be9b07.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/liballoc_cortex_m-902a7616d29b0ddc.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/liblinked_list_allocator-a436f22cf48dfc61.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libcortex_m-9a6749038f6e8315.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libembedded_hal-be2cd19e255d0054.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libvoid-c3ac384dc77c7432.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libnb-44e1f3fb8154762e.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libnb-d2fd7374c8472436.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libvolatile_register-31350fb3fd26a60d.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libvcell-5928b0c31e98c32b.rlib" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/libbare_metal-2ebd85282e1bb3be.rlib" "/home/sanket1729/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/thumbv7m-none-eabi/lib/liballoc-b1f8a41d3c5b29df.rlib" "/home/sanket1729/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/thumbv7m-none-eabi/lib/librustc_std_workspace_core-ba626387a2e50982.rlib" "/home/sanket1729/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/thumbv7m-none-eabi/lib/libcore-bc639632082e340c.rlib" "/home/sanket1729/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/thumbv7m-none-eabi/lib/libcompiler_builtins-87b489c55aae9c9f.rlib" "-Bdynamic" "--eh-frame-hdr" "-znoexecstack" "-L" "/home/sanket1729/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/thumbv7m-none-eabi/lib" "-o" "/home/sanket1729/rust-miniscript/embedded/target/thumbv7m-none-eabi/debug/deps/embedded-46e018162ba49788" "--gc-sections" "-Tlink.x"
  = note: rust-lld: error: section '.rodata' will not fit in region 'FLASH': overflowed by 263 bytes
          rust-lld: error: section '.rodata' will not fit in region 'FLASH': overflowed by 274 bytes
          rust-lld: error: section '.rodata' will not fit in region 'FLASH': overflowed by 276 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 276 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 276 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 280 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 284 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 288 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 288 bytes
          rust-lld: error: section '.rodata' will not fit in region 'FLASH': overflowed by 263 bytes
          rust-lld: error: section '.rodata' will not fit in region 'FLASH': overflowed by 274 bytes
          rust-lld: error: section '.rodata' will not fit in region 'FLASH': overflowed by 276 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 276 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 276 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 280 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 284 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 288 bytes
          rust-lld: error: section '.data' will not fit in region 'FLASH': overflowed by 288 bytes
          rust-lld: error: section '.rodata' will not fit in region 'FLASH': overflowed by 263 bytes
          rust-lld: error: section '.rodata' will not fit in region 'FLASH': overflowed by 274 bytes
          rust-lld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
          

error: could not compile `embedded` due to previous error

File: src/lib.rs [hunk 2]
```diff
@@ -864,3 +880,80 @@ mod tests {
...
+    #[cfg(all(not(feature = "std"), not(test)))]
+    mod mutex {
+        use core::cell::{RefCell, RefMut};
+        use core::ops::{Deref, DerefMut};
+
+        pub type LockResult<Guard> = Result<Guard, ()>;
+
+        pub struct Mutex<T: ?Sized> {
+            inner: RefCell<T>,
+        }
```
[the remark this answers is not in the document]
Comment (sanket1729) [reply src/lib.rs:896]: The name `Mutex` might even be misleading if we cannot use its multi-threading environment. But I am glad the compiler can catch it.

At some point, if there is really a demand for it, spinlocks can be considered. https://docs.rs/spin/latest/spin/index.html seems like a good fit. 

Until anyone complains or demands this, we don't have to do anything. But, we should document this.

**Review by sanket1729 — APPROVED** [review]
ACK 28d304e88c58e0595643b2c60402c4839d92a28e

**Comment by sanket1729** [comment]
@mcroad, really sorry for the last time. This has conflicts again. After your next rebase, we can merge this and address the issues in follow-up PRs.

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mcroad) [reply src/lib.rs:900]: Documented the fact that `Mutex` is not a real mutex.

**Comment by mcroad** [comment]
Rebased and documented `no_std` `Mutex` stub.

**Review by sanket1729 — APPROVED** [review]
ACK ac16e05aa9943152e4235b490fd4ad03a496a7c1

**Comment by sanket1729** [comment]
Reviewed the range-diff. Merging this to avoid further conflicts. @mcroad, can you please clarify if the debug build is supposed to fail as mentioned in my comment above?

**Pull request merged**